<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tutorial · Crispulator.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="index.html"><img class="logo" src="assets/logo.png" alt="Crispulator.jl logo"/></a><h1>Crispulator.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search.html"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="index.html">Home</a></li><li class="current"><a class="toctext" href="tutorial.html">Tutorial</a><ul class="internal"><li><a class="toctext" href="#Graphical-overview-1">Graphical overview</a></li><li><a class="toctext" href="#Getting-started-1">Getting started</a></li><li><a class="toctext" href="#Running-simulation-1">Running simulation</a></li><li><a class="toctext" href="#Output-1">Output</a></li></ul></li><li><a class="toctext" href="custom.html">Custom Simulations</a></li><li><a class="toctext" href="internals.html">Simulation Internals</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href="tutorial.html">Tutorial</a></li></ul><a class="edit-page" href="https://github.com/tlnagy/Crispulator.jl/blob/master/docs/src/tutorial.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Tutorial</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Tutorial-1" href="#Tutorial-1">Tutorial</a></h1><p>The easiest way to use Crispulator is via the command line config file. If this format is too constraining, the <a href="custom.html#Custom-Simulations-1">Custom Simulations</a> has a detailed walk- through of writing a custom simulation where each step can be modified according to need.</p><h2><a class="nav-anchor" id="Graphical-overview-1" href="#Graphical-overview-1">Graphical overview</a></h2><p>The simulation is laid out in the following manner:</p><p><img src="assets/overview.png" alt/></p><h2><a class="nav-anchor" id="Getting-started-1" href="#Getting-started-1">Getting started</a></h2><p>First, navigate to the Crispulator directory.</p><div class="admonition tip"><div class="admonition-title">Tip</div><div class="admonition-text"><p>You can find the directory by running</p><pre><code class="language-sh">$ julia -e &#39;println(Pkg.dir(&quot;Crispulator&quot;))&#39;</code></pre></div></div><p>There should be a <a href="https://learnxinyminutes.com/docs/yaml/">YAML file</a> called <code>example_config.yml</code>. Open this is in a text editor and it should look like this</p><pre><code class="language-yaml"># This is an example configuration file. Whitespace is important.

# Settings pertaining to the library design
library:
    genome:
        num-genes: 500
        num-guides-per-gene: 5
        frac-increasing-genes: 0.02 # fraction of genes with a positive phenotype
        frac-decreasing-genes: 0.1 # fraction of genes with a negative phenotype

    guides:
        crispr-type: CRISPRn # either CRISPRi or CRISPRn
        frac-high-quality: 0.9 # fraction of high quality guides
        mean-high-quality-kd: 0.85 # mean knockdown by a high quality guide (CRISPRi only)

screen:
    type: facs # either facs or growth
    num-runs: 10 # how many independent runs

    representation: # integer value, how much larger are samples than the library
        - transfection: 100
        - selection: 100
        - sequencing: 100

# screen-type specific parameters

    bin-size: 0.25 # size of tail to sample from, must be between 0 and 0.5 (FACS only)
    std-noise: 1 # (FACS only)
    num-bottlenecks: 10 # (Growth only)</code></pre><p>This gives access to most dials in the simulation, if something is missing than see <a href="custom.html#Custom-Simulations-1">Custom Simulations</a>.</p><p>Now, lets remove all genes that have a positive phenotype by changing line 8 to 0.0:</p><pre><code class="language-yaml">        frac-increasing-genes: 0.0 # fraction of genes with a positive phenotype</code></pre><h2><a class="nav-anchor" id="Running-simulation-1" href="#Running-simulation-1">Running simulation</a></h2><p>Now, we can actually run the code by executing the following command</p><pre><code class="language-julia">julia src/run.jl config example_config.yml test_output</code></pre><div class="admonition tip"><div class="admonition-title">Tip</div><div class="admonition-text"><p>Here <code>config</code> tells <code>CRISPulator</code> to use the provided config <code>example_config.yml</code> and <code>test_output</code> is the directory where the results will be saved. This directory will be created if it doesn&#39;t exist.</p></div></div><p>The output should look like</p><pre><code class="language-">output = readstring(`julia ../../src/run.jl config example_config.yml test_output`) # hide
println(output) # hide</code></pre><p>The <code>test_output/</code> directory should now be populated with all the files</p><div><pre><code class="language-none">counts.svg
volcano.svg</code></pre></div><h2><a class="nav-anchor" id="Output-1" href="#Output-1">Output</a></h2><p>The folder contains one of the raw count scatterplots (left) and a volcano plot of mean log2 fold change versus significance of each gene (right)</p><p><img src="test_output/counts.svg" alt/> <img src="test_output/volcano.svg" alt/></p><p>It also has a useful table that contains all the summary statistic information.</p><pre><code class="language-">using DataFrames # hide
head(readtable(joinpath(&quot;test_output&quot;, &quot;results_table.csv&quot;))) # hide</code></pre><p>The table below describes each column</p><table><tr><th>Column Name</th><th>Meaning</th></tr><tr><td><code>method</code></td><td>Which summary statistic was used (e.g. <a href="internals.html#Simulation.auprc"><code>Simulation.auprc</code></a>)</td></tr><tr><td><code>measure</code></td><td>Whether the score is only for increasing genes (<code>inc</code>), decreasing (<code>dec</code>) or both (<code>incdec</code>). Allows independent evaluation on which type of genes the screen can accurately evaluate.</td></tr><tr><td><code>genetype</code></td><td>Whether the score is for linear, sigmoidal, or all genes (see <a href="internals.html#Simulation.KDPhenotypeRelationship"><code>Simulation.KDPhenotypeRelationship</code></a>). Helps determine if CRISPRn or CRISPRi is better for this design.</td></tr><tr><td><code>mean_score</code></td><td>Average score</td></tr><tr><td><code>std_score</code></td><td>Standard deviation in scores</td></tr><tr><td><code>conf_max</code></td><td>Upper limit of 95% confidence interval</td></tr><tr><td><code>conf_min</code></td><td>Lower limit of 95% confidence interval</td></tr><tr><td><code>n</code></td><td>Number of independent replicates</td></tr></table><footer><hr/><a class="previous" href="index.html"><span class="direction">Previous</span><span class="title">Home</span></a><a class="next" href="custom.html"><span class="direction">Next</span><span class="title">Custom Simulations</span></a></footer></article></body></html>
